File: SubscriptionApi/SubscriptionApi/Commands/SubscriptionStartCommandFromApi.cs
using System.Collections.Generic;
using SubscriptionApi.Dto;
using SubscriptionApi.Exceptions;
using SubscriptionApi.Queries;
using SubscriptionApi.ResponseBuilders;
using SubscriptionApi.Subscriptions;

namespace SubscriptionApi.Commands
{
  public interface SubscriptionStartCommand
  {
    void ValidateData();
    void Resolve();
    void Authorize();
    void Invoke();
    StartSubscriptionResponseDto Response();
  }

  public class SubscriptionStartCommandFromApi : SubscriptionStartCommand
  {
    private readonly NewSubscriptionParametersDto _parameters;
    private readonly List<string> _requestedAssetNames = new List<string>();
    private readonly AssetAccessAuthorization _authorizationStructure;
    private readonly SubscriptionStartResponseBuilder _responseBuilder;
    private readonly ISubscriptionFactory _subscriptionFactory;
    private readonly List<AssetQuery> _assetQueries = new List<AssetQuery>();
    private readonly SubscriptionsModifyOperations _subscriptions;
    

    public SubscriptionStartCommandFromApi(
      NewSubscriptionParametersDto parameters, 
      AssetAccessAuthorization authorizationStructure, 
      SubscriptionStartResponseBuilder responseBuilder, 
      ISubscriptionFactory subscriptionFactory, 
      SubscriptionsModifyOperations subscriptions, 
      List<AssetQuery> assetQueries)
    {
      _parameters = parameters;
      _authorizationStructure = authorizationStructure;
      _responseBuilder = responseBuilder;
      _subscriptionFactory = subscriptionFactory;
      _subscriptions = subscriptions;
      _assetQueries = assetQueries;
    }

    public void ValidateData()
    {
      //TODO user name
      //TODO duration
      //TODO validate each query

      throw new System.NotImplementedException();
    }

    public void Resolve()
    {
      foreach (var query in _assetQueries)
      {
        query.ResolveInto(_requestedAssetNames, _responseBuilder);
      }

      _responseBuilder.AssertNoFatalErrors(new FatalErrorOccuredDuringQueryResolutionException());
    }

    public void Authorize()
    {
      foreach (var assetName in _requestedAssetNames)
      {
        _authorizationStructure.VerifyAccessTo(
          assetName, 
          _parameters.UserName, 
          _responseBuilder);
      }

      _responseBuilder.AssertNoFatalErrors(
        new FatalErrorOccuredDuringAuthorizationException());
    }

    public void Invoke()
    {
      var subscription = _subscriptionFactory.CreateFrom(_parameters);
      _subscriptions.AddNew(subscription);
    }


    public StartSubscriptionResponseDto Response()
    {
      return _responseBuilder.Build();
    }
  }

  //Add Subscription.SendNotification()
}

File: Java/factories/_01_no_factories/src/main/java/Commands/SubscriptionStartCommandFromApi.java
package commands;

import dto.NewSubscriptionParametersDto;
import dto.StartSubscriptionResponseDto;
import exceptions.FatalErrorOccuredDuringAuthorizationException;
import exceptions.FatalErrorOccuredDuringQueryResolutionException;
import queries.AssetQuery;
import responseBuilders.SubscriptionStartResponseBuilder;
import subscriptions.ISubscriptionFactory;
import subscriptions.Subscription;
import subscriptions.SubscriptionsModifyOperations;

import java.util.ArrayList;
import java.util.List;

public class SubscriptionStartCommandFromApi implements SubscriptionCommand {
    private final NewSubscriptionParametersDto parametersDto;
    private final List<String> _requestedAssetNames = new ArrayList<>();
    private final AssetAccessAuthorization authorizationStructure;
    private final SubscriptionStartResponseBuilder responseBuilder;
    private final ISubscriptionFactory subscriptionFactory;
    private final List<AssetQuery> assetQueries;
    private final SubscriptionsModifyOperations subscriptions;


    public SubscriptionStartCommandFromApi(
        NewSubscriptionParametersDto parameters,
        AssetAccessAuthorization authorizationStructure,
        SubscriptionStartResponseBuilder responseBuilder,
        ISubscriptionFactory subscriptionFactory,
        SubscriptionsModifyOperations subscriptions,
        List<AssetQuery> assetQueries) {

        this.parametersDto = parameters;
        this.authorizationStructure = authorizationStructure;
        this.responseBuilder = responseBuilder;
        this.subscriptionFactory = subscriptionFactory;
        this.subscriptions = subscriptions;
        this.assetQueries = assetQueries;
    }

    public void validateData() {
        //TODO user name
        //TODO duration
        //TODO validate each query

        throw new RuntimeException("not implemented");
    }

    public void resolve() {
        for (AssetQuery query : assetQueries) {
            query.resolveInto(_requestedAssetNames, responseBuilder);
        }

        responseBuilder.assertNoFatalErrors(
            new FatalErrorOccuredDuringQueryResolutionException());
    }

    public void authorize() {
        for (String assetName : _requestedAssetNames) {
            authorizationStructure.verifyAccessTo(
                assetName,
                parametersDto.userName,
                responseBuilder);
        }

        responseBuilder.assertNoFatalErrors(
            new FatalErrorOccuredDuringAuthorizationException());
    }

    public void invoke() {
        Subscription subscription = subscriptionFactory.createFrom(parametersDto);
        subscriptions.addNew(subscription);
    }


    public StartSubscriptionResponseDto response() {
        return responseBuilder.buildStart();
    }
}
